### ISSUE TITLE
False positive `unused_variables` when the variable is used in conditional code

### ISSUE THREAD

**Issue by daira** [rust-lang/rust#147836]
### Summary

Consider code like this (simplified from real code [here](https://github.com/zcash/librustzcash/pull/1999/files#diff-57d91750cf8573e5cfebdee8bd31b2d635cb02adf7f5e5fdd20ff2253e43782fR1037-R1055)):

```rust
if let Some(v) = ... {
    // side effects

    #[cfg(feature = "some-feature")]
    code_that_uses(v);
}
```

It is not unexpected that this causes an `unused_variables` warning when "some-feature" is disabled, but I think it shouldn't. In fact this is quite a common pattern and is almost never a real bug, in my experience. Note that, even if `v` isn't used in a particular configuration, we must name it something because it is used in *some* configurations, and in any case the `if let Some` can't be omitted.

There are two ways to suppress the warning.
* We can call the variable by a name starting `_`. But I find that ugly: I prefer to reserve `_` for unused parameters, because that genuinely is a little unusual and deserves to be called out, whereas having a variable that is only used in conditional code is (at least for the way we use features in my workplace) common.
* We can add `#[allow(unused_variables)]`. But that is verbose and arguably shouldn't be needed. The variable *is* used. It is not likely to be an error that it isn't *always* used &mdash; any more than it would be if we had written `if (some_feature) { code_that_uses(v); }`, which doesn't and shouldn't produce a warning.

Both approaches risk a false negative under maintenance, i.e. if I inadvertently removed the use of `v` in the conditional code (which might be less obvious if there is more distance between the declaration and use).

### Lint Name

`unused_variables`

### Reproducer

See above.

### Version

I saw it in ~~`clippy 0.1.81`~~ `rustc 1.81.0` but I'm pretty sure it still does this in more recent versions.

**Comment by daira** [comment]
A related case (let me know if I should file this as a separate issue) is `unused_mut` when the mutation is only in conditional code. This also seems to come up fairly frequently for me.

**Comment by cjgillot** [comment]
This is not a bug, just a surprising way two features interact.

`#[cfg(stuff)]` and `if cfg!(stuff) {` behave very differently. When `stuff` is false, `if cfg!(sutff) {` just becomes a `if false {`. Meanwhile `#[cfg(stuff)]` **removes the code**, which is useful if it would be ill-typed without the feature.

In this case, we expect the `unused_variable` lint to fire: the only use *has been removed from code* right after macro expansion.
